internals/garbage-collector.rst

```diff
@@ -356,37 +356,68 @@ follows these steps in order:
    the reference counts fall to 0, triggering the destruction of all unreachable
    objects.
 
-Optimization: generations
-=========================
+Optimization: incremental collection
+====================================
 
 In order to limit the time each garbage collection takes, the GC
-implementation for the default build uses a popular optimization:
-generations. The main idea behind this concept is the assumption that most
-objects have a very short lifespan and can thus be collected soon after their
-creation. This has proven to be very close to the reality of many Python
+implementation for the default build uses incremental collection with two
+generations.
+
+The purpose of generations is to take advantage of what is known as the weak
+generational hypothesis: Most objects die young.
+This has proven to be very close to the reality of many Python
 programs as many temporary objects are created and destroyed very quickly.
 
 To take advantage of this fact, all container objects are segregated into
-three spaces/generations. Every new
-object starts in the first generation (generation 0). The previous algorithm is
-executed only over the objects of a particular generation and if an object
-survives a collection of its generation it will be moved to the next one
-(generation 1), where it will be surveyed for collection less often. If
-the same object survives another GC round in this new generation (generation 1)
-it will be moved to the last generation (generation 2) where it will be
-surveyed the least often.
-
-The GC implementation for the free-threaded build does not use multiple
-generations.  Every collection operates on the entire heap.
+two generations: young and old. Every new object starts in the young generation.
+
+
+To collect all unreachable cycles in the heap, the garbage collector must scan the
+whole heap. This whole heap scan is called a cycle.
+
+In order to limit the time each garbage collection takes, the previous algorithm
+is executed only on a portion of the heap called an increment. For each cycle,
+the increments will cover the whole heap.
+
+Each increment, the portion of the heap scanned by a single collection is made up
+of three parts:
+
+* The young generation
+* The oldest fraction of the old generation
+* All any objects reachable from those objects that have not yet been scanned this cycle.
+
+Any objects surviving this collection are moved to the old generation.
+ollection from cycles.
+
+When a cycle starts, no objects in the heap are considered to have been scanned.
+When all objects in the heap have been scanned a cycle ends, and all objects are
+considered unscanned again.
+
+In order to collect all unreachable cycles, each increment must contain all of
```

Comment:
To make sure that I understand this section on unreachable cycles, I think you're saying that we want to ensure that we fully capture the unreachable cycle because we want to ensure that the cycle is either fully gc'd or not, to avoid partial processing, which could be problematic later on?

If so, maybe it's worth mentioning explicitly.

Reply:
I think this could benefit from some concrete simple example either in English or with ascii art or some diagrams or some pseudo code.